Extending Opaque Types (#853)
fzyzcjy/flutter_rust_bridge: book/src/feature/lang_opaque.md

# Arbitrary types (opaque)

> Any Rust type, even if it is not supported using features of this library, can be used in Dart. This is done by wrapping it with `RustOpaque` also Rust can get opaque Dart type in the `DartOpaque` wrap.

Reviewer: ```suggestion
Any Rust type, even if it is not supported using features of this library, can be used in Dart. This is done by wrapping it with `RustOpaque`. In addition, Rust can get opaque Dart type in the `DartOpaque` wrap.
```

> Any Rust type, even if it is not supported using features of this library, can be used in Dart. This is done by wrapping it with `RustOpaque`. In addition, Rust can get opaque Dart type in the `DartOpaque` wrap.

Reviewer: ```suggestion
On one hand, any Rust type, even if it is not supported using features of this library, can be used in Dart. This is done by wrapping it with `RustOpaque`.

On the other hand, Rust can get arbitrary Dart type as well using the `DartOpaque` opaque wrapper.
```

> The Dart rust opaque objects should be disposed *manually*, though it will also be disposed when it is GCed, that is discouraged, due to [suggestions by Dart team](https://github.com/fzyzcjy/flutter_rust_bridge/issues/775#issuecomment-1274635037). Think of it just like a lot of Flutter objects that we are familiar with, such as `ui.Image` - we have to manually dispose them as well. The Dart opaque objects are always deleted on the parent Dart thread.

Reviewer: Hmm what is "dart rust opaque" that should be disposed manually? Looks like a typo

> The Rust opaque objects in Dart should be disposed *manually*, though it will also be disposed when it is GCed, that is discouraged, due to [suggestions by Dart team](https://github.com/fzyzcjy/flutter_rust_bridge/issues/775#issuecomment-1274635037). Think of it just like a lot of Flutter objects that we are familiar with, such as `ui.Image` - we have to manually dispose them as well. The Dart opaque objects are always deleted on the parent Dart thread.

Reviewer: uber-nit: Shall we split DartOpaque and RustOpaque into two separate files (webpages)

opaque.dispose();
```

## Example `DartOpaque` 

Rust:

```rust,noplayground
pub fn get_dart_opaque(a: DartOpaque) { ... }
pub fn return_dart_opaque() -> DartOpaque { ... }
...
```

And use it in Dart:

```dart
var opaque = await api.getDartOpaque(() => '42');
Object answer_obj = await api.returnDartOpaque();
var fn = answer_obj as String Function();
print(fn());
```

## Implementation details

> As for how it is implemented as well as the design towards safety, please refer to [this doc](../contributing/rust_opaque_type_safety.md).
> [this doc](../contributing/dart_opaque_type_safety.md).

Reviewer: ```suggestion
As for how it is implemented as well as the design towards safety, please refer to [this doc](../contributing/rust_opaque_type_safety.md) and [this doc](../contributing/dart_opaque_type_safety.md).
```
